core/src/impl/Kokkos_Utilities.hpp

```diff
@@ -143,6 +143,64 @@ struct destruct_delete {
 template <class...>
 struct type_list;
 
+//------------------------------------------------------------------------------
+// <editor-fold desc="type_list_remove_first"> {{{2
+
+// Currently linear complexity; if we use this a lot, maybe make it better?
+
+template <class Entry, class InList, class OutList>
+struct _type_list_remove_first_impl;
+
+template <class Entry, class T, class... Ts, class... OutTs>
+struct _type_list_remove_first_impl<Entry, type_list<T, Ts...>,
+                                    type_list<OutTs...>>
+    : _type_list_remove_first_impl<Entry, type_list<Ts...>,
+                                   type_list<OutTs..., T>> {};
+
+template <class Entry, class... Ts, class... OutTs>
+struct _type_list_remove_first_impl<Entry, type_list<Entry, Ts...>,
+                                    type_list<OutTs...>>
+    : _type_list_remove_first_impl<Entry, type_list<>,
+                                   type_list<OutTs..., Ts...>> {};
+
+template <class Entry, class... OutTs>
+struct _type_list_remove_first_impl<Entry, type_list<>, type_list<OutTs...>>
+    : identity<type_list<OutTs...>> {};
+
+template <class Entry, class List>
```

Comment:
`List` must be `type_list<...>`, isn't it? Can we assert that?

Reply:
The usual way of asserting that is to assert this is to make the template undefined for specializations that are not a `type_list`. We do this for the implementation template, `_type_list_remove_first_impl`. We could also do this for `type_list_remove_first` (at the cost of a little additional compilation time, but probably not much) if you want. Alternatively, we can make it work for anything by doing something like:

```c++
template <class Entry, class List>
struct type_list_remove_first;

template <class Entry, template <class...> class Container, class... Items>
struct type_list_remove_first<Entry, Container<Items...>>
  : /* ... */ {};
```

but I expect this would cost significantly more compilation time without providing much benefit currently.

Reply:
I think I like something like
```C++
template <class T>
struct is_type_list : std::false_type {};

template <class... T>
struct is_type_list<type_list<T...>> : std::true_type {};

template <class Entry, class List>
struct type_list_remove_first
{
    static_assert(is_type_list<List>::value, "List must be of type type_list<...>!");
};

template <class Entry, class... Items>
struct type_list_remove_first<Entry, type_list<Items...>>
    : _type_list_remove_first_impl<Entry, type_list<Items...>, type_list<>> {};
```
since it gives a pretty clear error message. What do you think?